File: webapp/lib/ability_effect.rb
# frozen_string_literal: true

# Interprets an ability card's declarative `effect` spec from cards.yaml. As
# with boss and room effects, the behaviour is data. The shape (each optional):
#
#   effect:
#     add_damage: 2        # +N to the targeted room this crawl (Reinforcements)
#     unreducible: true    # the targeted room's damage cannot be reduced (Expose Weakness)
#     zero: true           # the targeted room deals 0 this crawl (Sabotage)
#     retreat: true        # the party turns back at the targeted room (Retreat):
#                          #   rooms before it still resolve, the rest are skipped
#     draw_rooms: 2        # the player draws N room cards (Blueprints)
#
# `add_damage`, `unreducible`, `zero`, and `retreat` act on a chosen room, so
# those cards need a room target; `draw_rooms` does not.
module AbilityEffect
  class Spec
    def initialize(raw = {})
      raw = raw.is_a?(Hash) ? raw : {}
      @add_damage  = raw["add_damage"] && Integer(raw["add_damage"])
      @unreducible = raw["unreducible"] ? true : false
      @zero        = raw["zero"] ? true : false
      @retreat     = raw["retreat"] ? true : false
      @draw_rooms  = raw["draw_rooms"] && Integer(raw["draw_rooms"])
    end

    attr_reader :add_damage, :draw_rooms

    def unreducible?
      @unreducible
    end

    def zero?
      @zero
    end

    def retreat?
      @retreat
    end

    # Whether this card must be played on a specific room.
    def targets_room?
      !@add_damage.nil? || @unreducible || @zero || @retreat
    end
  end

  def self.for(card)
    Spec.new(card.respond_to?(:effect) ? card.effect : {})
  end
end

File: android/app/src/main/java/com/dungeonboss/game/AbilityEffect.kt
package com.dungeonboss.game

import com.dungeonboss.model.AbilityCard

/**
 * Interprets an ability card's declarative `effect` spec (see docs/cards.md).
 * `add_damage`, `unreducible`, `zero`, and `retreat` act on a chosen room, so
 * those cards need a room target; `draw_rooms` does not. Mirrors
 * `webapp/lib/ability_effect.rb`.
 */
object AbilityEffect {
    class Spec(raw: Map<String, Any?>) {
        /** +N to the targeted room this crawl, or null. */
        val addDamage: Int? = if (raw.containsKey("add_damage")) Effects.intOf(raw["add_damage"]) else null

        /** Number of room cards to draw, or null. */
        val drawRooms: Int? = if (raw.containsKey("draw_rooms")) Effects.intOf(raw["draw_rooms"]) else null

        val unreducible: Boolean = Effects.boolOf(raw["unreducible"])
        val zero: Boolean = Effects.boolOf(raw["zero"])
        val retreat: Boolean = Effects.boolOf(raw["retreat"])

        /** Whether this card must be played on a specific room. */
        fun targetsRoom(): Boolean = addDamage != null || unreducible || zero || retreat
    }

    fun forCard(card: AbilityCard): Spec = Spec(card.effect)
}
